### PR TITLE
Use a function `eval_poly` for polynomial evaluation

### LINKED ISSUE #167: Investigate using `np.polynomial.Polynomial`
Over in #69, @tztsai pointed out `np.poly_1d` as a way to simplify code - and although `np.polynomial.Polynomial` is now the preferred interface - that looks like it could be both a speed up and a way to simplify the code for many polynomials in the code. 

A simple `timeit` comparison shows a significant speedup, but worth checking how it scales. Interestingly it looks like the overhead in creating a `Polynomial` instance isn't negligible, but I don't think it's worth creating `Polynomial` instances in the underlying constants classes.

```python

In [25]: import timeit
In [26]: m1 = timeit.timeit(
    ...:     stmt="""po = 0.99983952 + (6.788260e-5) * tc
    ...: po += -(9.08659e-6) * tc * tc
    ...: po += (1.022130e-7) * tc * tc * tc
    ...: po += -(1.35439e-9) * tc * tc * tc * tc
    ...: po += (1.471150e-11) * tc * tc * tc * tc * tc
    ...: po += -(1.11663e-13) * tc * tc * tc * tc * tc * tc
    ...: po += (5.044070e-16) * tc * tc * tc * tc * tc * tc * tc
    ...: po += -(1.00659e-18) * tc * tc * tc * tc * tc * tc * tc * tc
    ...: """,
    ...:     setup="""import numpy as np
    ...: tc = np.array([10,20,30])
    ...: """,
    ...: )
In [27]: m2 = timeit.timeit(
    ...:     stmt="""pn = np.polynomial.Polynomial(cf)
    ...: po = pn(tc)
    ...: """,
    ...:     setup="""import numpy as np
    ...: tc = np.array([10,20,30])
    ...: cf = np.array([0.99983952 , 6.788260e-5 , -9.08659e-6 , 1.022130e-7 , -1.35439e-9 ,
    ...:                1.471150e-11, -1.11663e-13, 5.044070e-16, -1.00659e-18])
    ...: """,
    ...: )
In [28]: m3 = timeit.timeit(
    ...:     stmt="po = pn(tc)",
    ...:     setup="""import numpy as np
    ...: tc = np.array([10,20,30])
    ...: cf = np.array([0.99983952 , 6.788260e-5 , -9.08659e-6 , 1.022130e-7 , -1.35439e-9 ,
    ...:                1.471150e-11, -1.11663e-13, 5.044070e-16, -1.00659e-18])
    ...: pn = np.polynomial.Polynomial(cf)
    ...: """,
    ...: )
In [29]: print(m1, m2, m3)
40.16773283301154 29.03079683400574 23.08425579199684
```

_Originally posted by @davidorme in https://github.com/ImperialCollegeLondon/pyrealm/pull/69#discussion_r1450519580_

### PR DESCRIPTION
**Pull request by tztsai**
# Description

A function `eval_poly` is implemented in `pyrealm.core.utilities` using Horner's method for polynomial evaluation and used to simplify H2O density calculation and specific heat calculation of air.

Fixes #167 

## Type of change

- [x] Optimization (back-end change that speeds up the code)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pyrealm/core/utilities.py [hunk 1]
```diff
@@ -336,3 +336,17 @@ def bounds_mask(
     )
 
     return outputs
+
+
+def eval_poly(x: NDArray, cf: Union[list, NDArray]) -> NDArray:
+    """Evaluates a polynomial with coefficients cf at the values in x.
```
[the remark this answers is not in the document]
Comment (davidorme) [reply pyrealm/core/utilities.py:342]: And maybe a slightly longer name that gives that info (`evaluate_horner_polynomial`)?

[the remark this answers is not in the document]
Comment (tztsai) [reply pyrealm/core/utilities.py]: I have addressed these now

**Review by davidorme — CHANGES_REQUESTED** [review]
I've pushed a minor change to the docstring to use math notation (and to show the change in equation form, rather than a specific evaluation) but otherwise this looks great.

I forgot one thing - can you add some simple unit tests? It doesn't need to be complex - a few example sets of coefficients and some $x$ values that can then be assessed using the 'normal' form and the new function? Something with a few different parameter sets that basically does this:

```python
import numpy as np
from pyrealm.core.utilities import evaluate_horner_polynomial

cf = np.array([5,4,3,2])
x = 2

normal = np.sum(cf * np.power(x, range(len(cf))))
horner = evaluate_horner_polynomial(x, cf)
assert normal == horner
```

File: tests/unit/core/test_evaluate_horner_polynomial.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+"""Test of the evaluate_horner_polynomial function."""
+
+import numpy as np
+
+from pyrealm.core.utilities import evaluate_horner_polynomial
+
+
+def test_evaluate_horner_polynomial():
+    """Test the evaluate_horner_polynomial function."""
+    for deg in range(10):
```
Comment (davidorme) [inline tests/unit/core/test_evaluate_horner_polynomial.py:10]: This is being super picky but it's good to have a bit more context in unit test.

```suggestion
    """Test the evaluate_horner_polynomial function.
    
    This checks that the normal form (a + bx + cx2 + dx3) and output from the
    horner form evaluation (a + x(b + x(c + dx))) are equivalent for powers up to 10.
    """
    for deg in range(10):
```

File: tests/unit/core/test_evaluate_horner_polynomial.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+        cf = np.random.rand(deg + 1)
+        x = np.random.rand()
```
Comment (davidorme) [inline tests/unit/core/test_evaluate_horner_polynomial.py:12]: Mostly just to check it all works with negative values 😄 Could equally use a uniform distribution in [-1, 1].

```suggestion
        cf = np.random.normal(deg + 1)
        x = np.random.normal()
```

**Review by davidorme — CHANGES_REQUESTED** [review]
LGTM - some minor changes

**Comment by davidorme** [comment]
I fixed the `qa` failure - somehow a `black` issue had snuck past pre-commit. The tests are failing but I think the `docs` failure is a VM fail - it runs on the `push` but not on the `pr` event.

**Comment by tztsai** [comment]
> I've pushed an update of the failing test and expanded it a bit to make sure it handles different inputs. Can you make sure you run `pytest` before pushing commits? Particularly when working on tests, making sure that things run locally before setting off GitHub Actions can save a lot of waiting around!

Thanks for the update!

**Pull request merged**